### DISCUSSION TITLE
Cara menyimpan DB di shiny

### DISCUSSION THREAD

**Discussion by rafiff23** [teamalgoritma/community#621] [Q&A] [answered]
Saya menggunakan ShinyManager untuk membuat login page codingan saya persis seperti yang ada disini https://datastorm-open.github.io/shinymanager/ yang disarankan oleh mas @Yusraf002 
dari diskusi saya sebelumnya di https://github.com/teamalgoritma/community/discussions/614 saya sudah mengganti passphrase nya dan saat di upload ke shiny, codingan saya berjalan dengan lancar, tetapi ada yang janggal karena setiap shiny server mati atau idle, hasil data yang ada di DB sebelumnya hilang apabila memulai server kembali. Dan ternyata memang di website shinymanager mengatakan seperti itu juga, jadi bagaimana caranya membuat DB tersebut tidak hilang selama lamanya apakah harus menggunakan postgresql? jika iya apakah bisa dijelaskan step yang harus dilakukan untuk mengconnect postgresql dan apakah codingan harus diubah juga dari sbeelumnya

**Comment by Yusraf002** [comment, marked as answer]
Dear @rafiff23 ,
Terimakasih atas pertanyaannya pak. pernyataan yang bapak sampaikan sudah betul pak, agar data kita tidak hilang kita perlu menghubungkan dengan PostgreSQL. berikut contoh untuk menghubungkan koding yang sudah bapak buat diatas dengan PostgreSQL.
```{r}
library(shiny)
library(shinymanager)
library(RPostgres)

# Connect to the PostgreSQL database
con <- dbConnect(
  Postgres(),
  user = "username",
  password = "password",
  host = "localhost",
  port = 5432,
  dbname = "database_name"
)

# Create a table in the database
dbExecute(con, "CREATE TABLE credentials (id SERIAL, user TEXT, password TEXT, admin BOOLEAN)")

# Insert data into the table
dbExecute(con, "INSERT INTO credentials (user, password, admin) VALUES ('user', 'user', FALSE), ('Adtree', '270792', TRUE)")

# Define UI
ui <- secure_app(
  # App UI code here
)

# Define server
server <- function(input, output, session) {
  
  # Authenticate users
  res_auth <- secure_server(
    check_credentials = check_credentials(
      con,
      table_name = "credentials",
      username_col = "user",
      password_col = "password",
      salt = "somesalt"
    )
  )
  
  # Your server code here
  
}

# Run the app
shinyApp(ui, server)

# Disconnect from the database
dbDisconnect(con)
```

untuk lebih jelasnya bapak bisa pelajari lebih lanjut mengenai menghubungkan rshiny dengan postgrsql melalui link berikut:
https://solutions.posit.co/connections/db/databases/postgresql/

Sekian informasi dari kami, jika dirasa membantu dapat klik "mark as answer" pada comment ini. Terima kasih

Best Regards,
Algoritma Team.
